torchaudio/prototype/models/hifi_gan.py

```diff
@@ -0,0 +1,307 @@
+        resblock_dilation_sizes (tuple of tuples of int): Dilation sizes for each 1D convolutional layer in each
+        residual block. For resblock type 1 inner tuples should have length ``3``, because there are ``3``
+            convolutions in each layer. For resblock type 2 they should have length 2.
+        resblock_type (int, 1 or 2): Determines whether ``ResBlock1`` or ``ResBlock2`` will be used.
+        lrelu_slope (float): Slope of leaky ReLUs in activations.
+    """
+
+    def __init__(
+        self,
+        in_channels: int,
+        upsample_rates: Tuple[int, ...],
+        upsample_initial_channel: int,
+        upsample_kernel_sizes: Tuple[int, ...],
+        resblock_kernel_sizes: Tuple[int, ...],
+        resblock_dilation_sizes: Tuple[Tuple[int, ...], ...],
+        resblock_type: int,
+        lrelu_slope: float,
+    ):
+        super(HiFiGANGenerator, self).__init__()
+        self.num_kernels = len(resblock_kernel_sizes)
+        self.num_upsamples = len(upsample_rates)
+        self.conv_pre = Conv1d(in_channels, upsample_initial_channel, 7, 1, padding=3)
+        resblock = ResBlock1 if resblock_type == 1 else ResBlock2
+
+        self.ups = nn.ModuleList()
+        for i, (u, k) in enumerate(zip(upsample_rates, upsample_kernel_sizes)):
+            self.ups.append(
+                ConvTranspose1d(
+                    upsample_initial_channel // (2**i),
+                    upsample_initial_channel // (2 ** (i + 1)),
+                    k,
+                    u,
+                    padding=(k - u) // 2,
+                )
+            )
+
+        self.resblocks = nn.ModuleList()
+        for i in range(len(self.ups)):
+            ch = upsample_initial_channel // (2 ** (i + 1))
+            for (k, d) in zip(resblock_kernel_sizes, resblock_dilation_sizes):
+                self.resblocks.append(resblock(ch, k, d, lrelu_slope))
+
+        self.conv_post = Conv1d(ch, 1, 7, 1, padding=3)
+        self.lrelu_slope = lrelu_slope
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        """
+        Args:
+            x (Tensor): Feature input tensor of shape (batch_size, num_channels, time_length)
+        Returns:
+            Tensor of shape ``(batch_size, 1, time_length * upsample_rate)``, where ``upsample_rate`` is the product
+            of upsample rates for all layers.
+        """
+        x = self.conv_pre(x)
+        for i, upsampling_layer in enumerate(self.ups):
+            x = F.leaky_relu(x, self.lrelu_slope)
+            x = upsampling_layer(x)
+            xs = torch.zeros_like(x)
+            for j in range(self.num_kernels):
+                res_block: ResBlockInterface = self.resblocks[i * self.num_kernels + j]
+                xs += res_block.forward(x)
+            x = xs / self.num_kernels
+
+        x = F.leaky_relu(x)
+        x = self.conv_post(x)
+        x = torch.tanh(x)
+
+        return x
+
+
+@torch.jit.interface
+class ResBlockInterface(torch.nn.Module):
```

Comment:
this is necessary to support torchscript, right?

Reply:
Yes, since variable `res_block`  in `HiFiGANGenerator.forward` is obtained by indexing `nn.ModuleList`, TorchScript requires it to have a type annotation. But it only recognizes a custom class as a type if I define it as an interface. `ResBlockInterface` has a docstring with a short explanation, I can expand it if needed